compiler/noirc_frontend/src/hir/mod.rs

```diff
@@ -119,18 +119,37 @@ impl Context {
         if &module_id.krate == crate_id {
             module_path
         } else {
-            let crate_name = &self.crate_graph[crate_id]
-                .dependencies
-                .iter()
-                .find_map(|dep| match dep {
-                    Dependency { name, crate_id } if crate_id == &module_id.krate => Some(name),
-                    _ => None,
-                })
-                .expect("The Struct was supposed to be defined in a dependency");
-            format!("{crate_name}::{module_path}")
+            let crates = self.find_dependencies(crate_id, &module_id.krate);
+            assert!(!crates.is_empty(), "The Struct was supposed to be defined in a dependency");
+            let mut fqn = String::new();
+            for name in crates {
+                fqn += &name.to_string();
+                fqn += "::";
+            }
+            fqn += &module_path;
+            fqn
         }
     }
 
+    /// Recursively walks down the crate dependency graph from crate_id until we reach requested crate
+    /// This is needed in case a library (lib1) re-export a structure defined in another library (lib2)
+    /// In that case, we will get [lib1,lib2] when looking for a struct defined in lib2, 
+    /// re-exported by lib1 and used by the main crate.
+    /// Returns the path from crate_id to target_crate_id
+    fn find_dependencies(&self, crate_id: &CrateId, target_crate_id: &CrateId) -> Vec<CrateName> {
+        for dep in &self.crate_graph[crate_id].dependencies {
+            if &dep.crate_id == target_crate_id {
+                return vec![dep.name.clone()];
+            }
+            let mut fqn = self.find_dependencies(&dep.crate_id, target_crate_id);
+            if !fqn.is_empty() {
+                fqn.insert(0, dep.name.clone());
+                return fqn;
+            }
+        }
+        Vec::new()
+    }
```

Comment:
```suggestion
    fn find_dependencies(&self, crate_id: &CrateId, target_crate_id: &CrateId) -> Option<Vec<String>> {
        for dep in &self.crate_graph[crate_id].dependencies {
            if &dep.crate_id == target_crate_id {
                return Some(vec![dep.name.clone()]);
            }
            
            if let Some(mut path) = self.find_dependencies(&dep.crate_id, target_crate_id) {
                path.insert(0, dep.name.to_string());
                return path;
            }
        }
        None
    }
```
- I think using an `Option` makes the failure case a bit clearer
- Returning Strings directly lets us use `join` when this function is called above.
- I tried changing to a VecDeque, but that doesn't support `join` unfortunately.

Reply:
Done